cupyx/scipy/sparse/linalg/solve.py

```diff
@@ -65,3 +66,266 @@ def lsqr(A, b):
+        tol (float): Tolerance for convergence, exits when
+                     residual norm(r) < tol * norm(b)
+        maxiter (int): maximum number of iterations (defaults to `numpy.inf`)
+        callback (callable): a callback function to call at each iteration
+                            (unlike scipy, it accepts iteration and
+                            residual as well as current vector `x`)
+        atol (float): Default to 0 (use tol instead of atol for convergence)
+
+    Returns:
+        tuple:
+            A tuple of result x and info (as in scipy, 0 if successful,
+            1 if no convergence, <0 if rho or omega breakdown)
+
+    .. seealso:: :func:`scipy.sparse.linalg.bicgstab`
+    """
+
+    if not cupyx.scipy.sparse.isspmatrix_csr(A):
+        A = cupyx.scipy.sparse.csr_matrix(A)
+    util._assert_nd_squareness(A)
+    util._assert_cupy_array(b)
+    n = A.shape[0]
+    if b.ndim != 1 or len(b) != n:
+        raise ValueError('b must be 1-d array whose size is same as A')
+
+    # support float32, float64, complex64, and complex128
+    if A.dtype.char in 'fdFD':
+        dtype = A.dtype.char
+    else:
+        dtype = numpy.promote_types(A.dtype.char, 'f').char
+
+    # Calculate initial residual
+    x0 = x0 if x0 is not None else cupy.zeros(n, dtype=dtype)
+    r = cupy.cusparse.spmv(A, x0, b, -1, 1)
+    p, q = cupy.zeros_like(r), cupy.zeros_like(r)
+    r_tilde = r.copy()
+
+    # Main BiCGSTAB loop
+
+    i = 0
+    rho = alpha = omega = cupy.linalg.norm(r)
+
+    info = 1
+
+    x = x0.copy()
+    tol = cupy.linalg.norm(b) * tol
+    atol = 0.0 if atol == 'legacy' else atol
+
+    # preconditioner
+    M, M_info = (_identity, None) if M is None else M
+
+    while info <= maxiter:
+        prev_rho = rho
+        rho = cupy.vdot(r_tilde, r)
+        if rho == 0:
+            info = -10
+            break
+        beta = rho / prev_rho * alpha / omega
+        p = r + beta * (p - omega * q)
+
+        # Alpha step
+        phat = M(p)  # apply preconditioner
+
+        q = cupy.cusparse.spmv(A, phat)
+        a = cupy.vdot(r_tilde, q)
+        if a == 0:
+            info = -11
+            break
+        alpha = rho / a
+        s = r - alpha * q
+
+        residual = cupy.linalg.norm(s)
+        if residual < tol or residual < atol:
+            x += alpha * phat
+            info = 0
+            break
+
+        # Omega step
+        shat = M(s)  # apply preconditioner
+
+        t = cupy.cusparse.spmv(A, shat)
+        omega = cupy.vdot(t, s) / cupy.square(cupy.linalg.norm(t))
+        if omega == 0:
+            info = -11
+            break
+
+        x += alpha * phat + omega * shat
+        r = s - omega * t
+
+        residual = cupy.linalg.norm(r)
+        if callback is not None:
+            callback(x, i, r)
+        if residual < tol or residual < atol:
+            info = 0
+            break
+        info += 1
+
+    _destroy_ilu(M_info)
```

Comment:
I was thinking this function will be useful for destroying `spilu` info in a more general context (in that case, I would remove the underscore), but if you think it isn't necessary (since `spilu` is meant to be primarily a preconditioner), I can inline it.